### PR TITLE
Fix branch matching for Buildkite

### PR DESCRIPTION
**Pull request by gsouquet**
Fixes vector-im/element-web#17719

Tested against @t3chguy failing PR by created a replica matrix-org/matrix-react-sdk#6235 / matrix-org/matrix-js-sdk#1751

### CODE REVIEW

File: scripts/fetchdep.sh [hunk 1]
```diff
@@ -25,30 +25,37 @@ clone() {
 # First we check if GITHUB_HEAD_REF is defined,
 # Then we check if BUILDKITE_BRANCH is defined,
 # if it isn't we can assume this is a Netlify build
-if [ -n ${GITHUB_HEAD_REF+x} ]; then
-	head=$GITHUB_HEAD_REF
-elif [ -n ${BUILDKITE_BRANCH+x} ]; then
-	head=$BUILDKITE_BRANCH
+if [ -z ${BUILDKITE_BRANCH+x} ]; then
```
Comment (jryans) [inline scripts/fetchdep.sh:28]: Hmm, at first glance, it seems a lot more confusing to follow when structuring each check as what is _not_ present... Why is that needed...? I would have though each env var only appeared on each service, i.e. `GITHUB_HEAD_REF` appears only on GHA and `BUILDKITE_BRANCH` appears only on Buildkite...?

File: scripts/fetchdep.sh [hunk 2]
```diff
@@ -25,30 +25,37 @@ clone() {
...
 else
-	# Netlify doesn't give us info about the fork so we have to get it from GitHub API
-	apiEndpoint="https://api.github.com/repos/matrix-org/matrix-react-sdk/pulls/"
-	apiEndpoint+=$REVIEW_ID
-	head=$(curl $apiEndpoint | jq -r '.head.label')
+	head=$BUILDKITE_BRANCH
 fi
 
-# If head is set, it will contain on BuilKite either:
+# If head is set, it will contain on BuildKite either:
```
Comment (jryans) [inline scripts/fetchdep.sh:41]: Buildkite uses a lowercase "K" in the name, even though internally we occasionally shorten it to "BK".

```suggestion
# If head is set, it will contain on Buildkite either:
```

File: scripts/fetchdep.sh [hunk 3]
```diff
@@ -25,30 +25,37 @@ clone() {
...
 #   * "branch" when the author's branch and target branch are in the same repo
 #   * "fork:branch" when the author's branch is in their fork or if this is a Netlify build
 # We can split on `:` into an array to check.
 # For GitHub Actions we need to inspect GITHUB_REPOSITORY and GITHUB_ACTOR
 # to determine whether the branch is from a fork or not
 BRANCH_ARRAY=(${head//:/ })
 if [[ "${#BRANCH_ARRAY[@]}" == "1" ]]; then
-    if [[ "$GITHUB_REPOSITORY" = "$deforg/$defrepo" ]]; then
-        clone $deforg $defrepo $head
+    if [ -z ${BUILDKITE_BRANCH+x} ]; then
```
Comment (jryans) [inline scripts/fetchdep.sh:49]: Let's try to rewrite this in terms of a positive check, i.e. test _for_ GitHub instead of "not Buildkite".

```suggestion
    if [ -n ${GITHUB_ACTIONS+x} ]; then
```

File: scripts/fetchdep.sh [hunk 4]
```diff
@@ -25,30 +25,37 @@ clone() {
...
+        if [[ "$GITHUB_REPOSITORY" == "$deforg"* ]]; then
```
Comment (jryans) [inline scripts/fetchdep.sh:50]: The docs suggest `GITHUB_REPOSITORY` is both owner and repo, so does that ever equal just the org...?

File: scripts/fetchdep.sh [hunk 5]
```diff
@@ -25,30 +25,37 @@ clone() {
...
+            clone $deforg $defrepo $GITHUB_HEAD_REF
+        else
+            clone $GITHUB_ACTOR $defrepo $GITHUB_HEAD_REF
```
Comment (jryans) [inline scripts/fetchdep.sh:53]: We may want to split the repo org out of `GITHUB_REPOSITORY`, rather than using `GITHUB_ACTOR`, since a manual workflow run might set `GITHUB_ACTOR` to something incorrect.

(This could be fixed separately if it's too much to worry about at the moment.)

**Review by jryans — CHANGES_REQUESTED** [review]
Thanks for looking into this. I'd like to see conditionals written positively before moving forward, as the negative version is more confusing to read.

File: scripts/fetchdep.sh (hunk 4 above)
Comment (gsouquet) [reply scripts/fetchdep.sh:50]: That's a check to see if `$GITHUB_REPOSITORY` starts with the value in `$deforg`. the `*` after `$deforg` acts as a wildcard

Comment (jryans) [reply scripts/fetchdep.sh:50]: Ah right, thanks, makes sense then. 😄

**Comment by gsouquet** [comment]
Thank you for the review @jryans I got tripped by some bash syntax and ended up writing some more convoluted code than what was required!
It should be in a much better place

File: scripts/fetchdep.sh [hunk 6]
```diff
@@ -25,38 +25,45 @@ clone() {
 # First we check if GITHUB_HEAD_REF is defined,
 # Then we check if BUILDKITE_BRANCH is defined,
```
[the remark this answers is not in the document]
Comment (jryans) [reply scripts/fetchdep.sh:26]: Perhaps it makes sense to keep GHA at the top though, like the comments currently say, since that's new home...?

**Review by jryans — CHANGES_REQUESTED** [review]
It's looking more readable overall, thanks! Let's sort out the conditionals at the top, then should be good to go.

File: scripts/fetchdep.sh [hunk 7]
```diff
@@ -25,10 +25,10 @@ clone() {
...
 # if it isn't we can assume this is a Netlify build
```
Comment (t3chguy) [inline scripts/fetchdep.sh:27]: if *they* aren't maybe?

**Review by jryans — APPROVED** [review]
Thanks, let's go for it! 😄
